### DISCUSSION TITLE
Typescript help with modular client extensions

### DISCUSSION THREAD

**Discussion by adamjkb** [prisma/prisma#18048] [Q&A]
I am currently in the process of developing a client extension for Prisma and as it grows I started to wonder what's the best way to modularize the extension's codebase while mainting typescript notations inside each fragment of code. Some pseudo code to represent what I mean:

```js
import * as methods from './methods/index.js'

export const myPrismaExtension = Prisma.defineExtension((client) => {
    return client.$extends({
        result: {
            page: {
                getChildren: methods.create_get_children_method(client),
                // ...and a lot more of these...
            }
        },
    })
})
```

And so inside of the `create_get_children_method()` I am not sure how to define the type of this function that returns the object of a `needs` object `compute` function to maintain the typing of the these to keys.

```js
/**
 * @param {import('@prisma/client').Prisma.DefaultPrismaClient} client
 */
export const create_get_children_method = (client) => ({
    needs: { id: true },
    compute: ({ id }) => {
        return asnyc () => client.page.findFirst()
    }
})
```

Obviously, I'm open to any suggestion on how to better break up a potentially complex client extensions. Breaking it to functions that defines the object seems like a good compromise since it gives me the ability to interact with the client and potentially pass in other options to each of these methods.

**Comment by ludralph** [comment]
Hi @adamjkb 

Prisma Client Extensions allow for modularization and type safety when extending the Prisma Client. You can define extensions in TypeScript, which should help you maintain type annotations throughout your codebase.

You can organize client extensions in a separate directory for maintainability. This implies that you can define your extensions in separate files and then import them into your main Prisma Client extension definition. This was also discussed in this [Github discussion](https://github.com/prisma/prisma/discussions/19725). The discussion also mentions the struggle to isolate types for a separated object, which seems similar to your concern about maintaining type safety in your `create_get_children_method` function.

The user in the discussion attempted to use the `Parameters` utility type to extract the types needed for their extensions. You might consider a similar approach or look into using `ReturnType` combined with `typeof` to capture the types of your extension methods.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
